spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/HealthCheckServiceInstanceListSupplier.java

```diff
@@ -56,50 +59,77 @@
-		defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
+		this.defaultHealthCheckPath = healthCheck.getPath().getOrDefault("default",
 				"/actuator/health");
 		this.webClient = webClient;
 		initInstances();
 
 	}
 
 	private void initInstances() {
-		delegate.get().subscribe(delegateInstances -> {
-			instances.clear();
-			instances.addAll(delegateInstances);
+		Flux<List<ServiceInstance>> instancesFlux = this.delegate.get().doOnNext(delegateInstances -> {
+			this.instances = Collections.unmodifiableList(new ArrayList<>(delegateInstances));
 		});
 
-		Flux<List<ServiceInstance>> healthCheckFlux = healthCheckFlux();
+		Flux<List<ServiceInstance>> verifiedInstancesFlux = healthCheckFlux()
+				.doOnNext((verifiedInstances -> {
+					this.healthyInstances = verifiedInstances;
+				}));
 
-		healthCheckFlux.subscribe(verifiedInstances -> {
-			healthyInstances.clear();
-			healthyInstances.addAll(verifiedInstances);
-		});
+		this.healthCheckDisposable = Flux.merge(instancesFlux.then(), verifiedInstancesFlux.then())
+				.subscribeOn(Schedulers.parallel())
+				.subscribe();
 	}
 
 	protected Flux<List<ServiceInstance>> healthCheckFlux() {
-		return Flux.create(emitter -> Schedulers
-				.newSingle("Health Check Verifier: " + getServiceId(), true)
-				.schedulePeriodically(() -> {
-					List<ServiceInstance> verifiedInstances = new ArrayList<>();
-					Flux.fromIterable(instances).filterWhen(this::isAlive)
-							.subscribe(serviceInstance -> {
-								verifiedInstances.add(serviceInstance);
-								emitter.next(verifiedInstances);
-							});
-				}, healthCheck.getInitialDelay(), healthCheck.getInterval().toMillis(),
-						TimeUnit.MILLISECONDS),
-				FluxSink.OverflowStrategy.LATEST);
+		return Flux.defer(() -> {
+			List<Mono<ServiceInstance>> checks = new ArrayList<>(this.instances.size());
+			for (ServiceInstance instance : this.instances) {
+				Mono<ServiceInstance> alive = isAlive(instance)
+						.onErrorResume(error -> {
+							if (LOG.isDebugEnabled()) {
+								LOG.debug(String.format(
+										"Exception occurred during health check of the instance for service %s: %s",
+										instance.getServiceId(), instance.getUri()), error);
+							}
+							return Mono.empty();
+						})
+						.timeout(this.healthCheck.getInterval(), Mono.fromSupplier(() -> {
+							if (LOG.isDebugEnabled()) {
+								LOG.debug(String.format(
+										"The instance for service %s: %s did not respond for %s during health check",
+										instance.getServiceId(), instance.getUri(), this.healthCheck.getInterval()));
+							}
+							//returns Completion signal if null
+							return null;
+						}))
+						.handle((isHealthy, sink) -> {
+							if (isHealthy) {
+								sink.next(instance);
+							}
+						});
+
+				checks.add(alive);
+			}
+			List<ServiceInstance> result = new CopyOnWriteArrayList<>();
+			return Flux.merge(checks).map(alive -> {
+				result.add(alive);
+				return result;
```

Comment:
as far as `List` is mutated should it be emitted on any update or only on first?  
this could be rewritten to  
```java
List<ServiceInstance> result = new CopyOnWriteArrayList<>();
AtomicBoolean first = new AtomicBoolean(false);
return Flux.merge(checks).<List<ServiceInstance>>handle((alive, sink) -> {
	result.add(alive);
	if (first.compareAndSet(false, true)) {
		sink.next(result);
	}
})
.defaultIfEmpty(result);
```

Reply:
Why not just `Flux.merge(checks).collectList().defaultIfEmpty(emptyList())`?

Reply:
this question was addressed to my intermediate implementation which mutated `healthyInstances` field.  
I was trying to preserve initial behavior.  
`.collectList()` will wait until all health checks are completed, in initial implementation `List` was emitted on the first healthy instance presence. As far as I understand it was made due to the fact that all health checks could take a long time (in current implementation it is `this.healthCheck.getInterval()` but in initial it could be `infinity`) but it should return healthy instances as soon as possible.